drivers/net/ethernet/microchip/enc28j60.c

```diff
@@ -1295,18 +1295,28 @@ static netdev_tx_t enc28j60_send_packet(struct sk_buff *skb,
 
 	/* Remember the skb for deferred processing */
 	priv->tx_skb = skb;
-	schedule_work(&priv->tx_work);
+	wake_up_process(priv->tx_thread);
 
 	return NETDEV_TX_OK;
 }
 
-static void enc28j60_tx_work_handler(struct work_struct *work)
+static int enc28j60_tx_thread(void *data)
 {
-	struct enc28j60_net *priv =
-		container_of(work, struct enc28j60_net, tx_work);
+	struct enc28j60_net *priv = (struct enc28j60_net *)data;
+
+	while (true) {
+		/* actual delivery of data */
+		if (priv->tx_skb)
+			enc28j60_hw_tx(priv);
+		set_current_state(TASK_IDLE);
```

Comment:
Please make this the first instruction in the loop. We should always set the state before we check for any value change in variables. Otherwise a similar race may happen for tx_skb as I pointed out for thread_stopped in my other comment to this function.